crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1370,25 +1370,24 @@ impl<'src> Lexer<'src> {
         // i.e., it recovered from an unclosed parenthesis (`(`, `[`, or `{`).
         self.nesting -= 1;
 
-        let current_position = self.current_range().start();
+        let mut current_position = self.current_range().start();
         let reverse_chars = self.source[..current_position.to_usize()].chars().rev();
-        let mut new_position = current_position;
-        let mut has_newline = false;
+        let mut newline_position = None;
 
         for ch in reverse_chars {
             if is_python_whitespace(ch) {
-                new_position -= ch.text_len();
+                current_position -= ch.text_len();
             } else if matches!(ch, '\n' | '\r') {
-                has_newline |= true;
-                new_position -= ch.text_len();
+                current_position -= ch.text_len();
+                newline_position = Some(current_position);
             } else {
                 break;
             }
         }
 
         // The lexer should only be moved if there's a newline character which needs to be
         // re-lexed.
-        if new_position != current_position && has_newline {
+        if let Some(newline_position) = newline_position {
             // Earlier we reduced the nesting level unconditionally. Now that we know the lexer's
```

Comment:
Now, we encode both information in a single variable:
1. The lexer needs to be moved
2. The position the lexer needs to be moved to